e2etest/declarativeRunner.go

```diff
@@ -33,15 +33,80 @@ import (
 // In particular, it lets one test cover a range of different source/dest types, and even cover both sync and copy.
 // See first test in zt_enumeration for an annotated example.
 
+var validCredTypesPerLocation = map[common.Location][]common.CredentialType{
+	common.ELocation.Unknown(): {common.ECredentialType.Unknown(), common.ECredentialType.Anonymous(), common.ECredentialType.OAuthToken()}, // Delete!
+	common.ELocation.File():    {common.ECredentialType.Anonymous()},
+	common.ELocation.Blob():    {common.ECredentialType.Anonymous(), common.ECredentialType.OAuthToken()},
+	common.ELocation.BlobFS():  {common.ECredentialType.Anonymous(), common.ECredentialType.OAuthToken()}, // todo: currently, account key auth isn't even supported in e2e tests.
+	common.ELocation.Local():   {common.ECredentialType.Anonymous()},
+	common.ELocation.Pipe():    {common.ECredentialType.Anonymous()},
+	common.ELocation.S3():      {common.ECredentialType.S3AccessKey()},
+	common.ELocation.GCP():     {common.ECredentialType.GoogleAppCredentials()},
+}
+
+var allCredentialTypes []common.CredentialType = nil
+var anonymousAuthOnly = []common.CredentialType{common.ECredentialType.Anonymous()}
+
+func getValidCredCombinationsForFromTo(fromTo common.FromTo, requestedCredentialTypesSrc, requestedCredentialTypesDst []common.CredentialType) [][2]common.CredentialType {
```

Comment:
I'm a bit confused with this implementation.
1. If we have `fromTo`, then `validCredTypesPerLocation[fromTo.from]` and `validCredTypesPerLocation[fromTo.To]` will give me respective lists of source and destination supported credTypes.
2. We'll just have to filter out all elements of `requestedCredentialTypesSrc` that area not supported. (`validCredTypesPerLocation[fromTo.from]`). Same step for destination.
3. Return back the filtered output but doing a cross product of both filtered lists.

Reply:
I... have questions about it myself. I wrote this almost a year ago now, so it's rather fuzzy in my head.

I *believe* the intent was to prevent a test author from inputting nonsensical auth types. I think the last bit where we directly check against _every_ credential type is not good code upon coming back to this. I'll rewrite this function a bit.

Reply:
Upon a bit more thought, the original implementation may be better. The set of possibilities available to either end of the from/to is always smaller than the requested set (assuming the requested set is allCredentialTypes).